### DISCUSSION TITLE
servo pulsing issue

### DISCUSSION THREAD

**Discussion by atomGit** [ExpressLRS/ExpressLRS#2876] [Q&A]
i don't know for certain that this is an ELRS issue, but it may be

this is a typical,, ground-based RC application - Radio, Receiver, servos - no FC

i'm driving 4 [GoBuilda 2000 series servos](https://www.servocity.com/2000-series-dual-mode-servo-25-2/) in continuous rotation mode and they all pulse, meaning that they change their speed, slightly, about 20-30 times for each rotation when stick input is around 80% or less - at near full stick the servos run smooth

i have an old HiTech servo programmer and when i drive one of the servos with that, it works fine, though it's 4.8v whereas i'm normally driving them with 7.4v - i don't know if that makes any difference

the Tx and Rx (8ch) are a RadioMaster Boxer, ELRS, both with the latest firmware, and i think i've tried messing with every radio and Rx setting that i thought might cause the problem, to no avail

i've contacted RadioMaster, ServoCity and have an open issue with [EdgeTX](https://github.com/EdgeTX/edgetx/discussions/5361)  and, so far, no one has been able to figure this out

Tx/Rx servo resolution, limits and update frequency seem to have no affect - no matter what i've tried, the problem still exists

ELRS added PWM output at some point ([PWM Servo Output Support · Issue #581](https://github.com/ExpressLRS/ExpressLRS/issues/581) and i'm using a ELRS PWM Rx ([ER8](https://www.radiomasterrc.com/products/er8-2-4ghz-elrs-pwm-receiver)) and so i'm wondering if there may be a conflict here???

in the elrsV3.lua script there are some items of potential interest, but i don't know what to try with them...

```
local elrsFlags = 0
local elrsFlagsInfo = ""
local deviceIsELRS_TX = nil
```

**Comment by pkendall64** [comment]
The ER8 is an ESP32 based receiver and the PWM signal are generated by the ESP32 chip itself using the MCPWM module that is built in. So the signals should be very stable and accurate. Do you have the ADC enabled in EdgeTX? Thats about the only thing I can think of.

**Comment by atomGit** [comment]
it was suggested to disable ADC but i tried it both ways

**Comment by atomGit** [comment]
i don't know if this will be of any use, but i noticed that the number of times the servo pulses (changes speed) seems to change with the % of stick input with the frequency lessening as stick input increases

**Comment by atomGit** [comment]
> So the signals should be very stable and accurate.

the servos run perfectly at any speed when driven by a HiTech servo programmer that's 20+ years old, so there must a difference with the PWM signal that's getting sent to them - what exactly is causing that, i'd sure like to know

**Comment by mha1** [comment]
Does it do the same if you don't control the servo with a stick but using a switch? I.e. define a mixer with a switch (instead of a stick) as input and set the weight to 70% or so to have a stick independent and constant source to drive the servo expecting a constant rotation speed.

**Comment by mha1** [comment]
And before trying a switch as input please try turning telemetry off. Start the ELRS LUA script and use line TLM ratio to select Off.

**Comment by atomGit** [comment]
> Does it do the same if you don't control the servo with a stick but using a switch?

yes, it does it with a switch, a knob, and the trims, with or without telm, on EdgeTX 2.8.4 and 2.10.2

**Comment by mha1** [comment]
does it change with packet rate (not to confuse with the servo update rate) changing it with the LUA script (first line packet rate) between 100Hz and 333Hz?

you mentioned 20-30 pulses per rotation. How long does a rotation roughly take?

**Comment by atomGit** [comment]
i tested all packet rates - no change

how long a rotation takes depends upon stick input, servo speed, etc. - with the current servos, ~1.5 seconds at full stick, maybe a bit less

**Comment by mha1** [comment]
I ran a 5s trace on an ER6 at 100Hz packet rate and 50Hz servo update rate with an EdgeTX mixer channel 6 set to output a fixed 70% which results in a nominal 1858us pulse length repeated every 20 ms.

![image](https://github.com/user-attachments/assets/33a2ef9f-20d1-4e78-89c7-cf913e77ead2)

![image](https://github.com/user-attachments/assets/ab714e1b-53e5-4965-a240-d2e1ae583adb)

If you calculate the pulse lengths from the traced duty period you can see the ER6 is really outputting the expected 1858us@50Hz. Doing this calculation for every pulse you can calculate the deviation for each pulse vs its expected value. The attached Excel report shows a maximum of 0.012us deviation with no regular pattern. 

The entire chain EdgeTX/ELRS/PWM output looks good to me.

I'd check the power supply and maybe the servo has a problem with 3.3V logic. Check if your servo tester outputs 5V PWM pulses. That might make a difference.  

[report.pdf](https://github.com/user-attachments/files/16494002/report.pdf)

**Comment by atomGit** [comment]
i appreciate you taking the time to test @mha1 :1st_place_medal: 

at around 70% channel output is where the PWM signal seems to start to smooth out - at 80-100% the servo sounds pretty normal to my ear

nevertheless, are we seeing these pulses in your data? there's 2 periods i'm curious about where deviation changes from the nominal 0,0072 - what's your opinion as to the severity of these deviations?

```
69 1363080940.00 9289964% 20.0 ms 1857,9928 0,0072
70 1383083600.00 9290064% 20.0 ms 1858,0128 -0,0128
71 1403086260.00 9290064% 20.0 ms 1858,0128 -0,0128
72 1423088920.00 9290055% 20.0 ms 1858,0110 -0,0110
73 1443091600.00 9289964% 20.0 ms 1857,9928 0,0072

155 3083309720.00 9289964% 20.0 ms 1857,9928 0,0072
156 3103312380.00 9290064% 20.0 ms 1858,0128 -0,0128
157 3123315040.00 9290064% 20.0 ms 1858,0128 -0,0128
158 3143317700.00 9290055% 20.0 ms 1858,0110 -0,0110
159 3163320380.00 9289964% 20.0 ms 1857,9928 0,0072
```

the time offset here is problematic - there's very little time (86 ns) between these deviations and i don't think this jives with what i'm seeing (~15 pulses / ~~sec.~~ servo rotation), however, if these deviations are creating a resonance by the time they reach the output shaft of a servo, well, i dunno

btw, the power supply for the Rx is 7.4vdc battery

i don't know how to test PWM voltage output of the servo tester - all i have is a multi-meter and i don't know how to use it for anything but the simplest of measurements

if you felt like it, i might suggest re-running your test for a longer period with channel output around 20% because somewhere around 70-80% is where the signal starts to smooth out

**Comment by mha1** [comment]
What do you mean by "the time offset here is problematic"?

Attached a 60s trace at 20% (1398us). I normalized the expected value to the median pulse length in this period. The deviation is in the range of [-18.6ns ; 1.4ns]. I think this is close to perfect. I doubt any servo will be able to act on 20ns changes meaning it would have a dead band less than 20ns. They are usually in the range of 1us.

[report 60s 20%.pdf](https://github.com/user-attachments/files/16501756/report.60s.20.pdf)

**Comment by atomGit** [comment]
thanks again for doing this

so what can be concluded? i assume we can conclude that ELRS is not an issue - is there anything else we can assume?

> What do you mean by "the time offset here is problematic"?

the period of time between those deviations i referenced don't match with the pulses i'm seeing/hearing ... in other words -- and given your 2nd set of data -- those deviations apparently had nothing to do with the servo pulsing issue

so if the problem isn't ELRS, that leaves the servos, the radio, the receiver, and EdgeTX - not a good situation, but at least i'm comfortable it's not an ELRS issue

**Comment by mha1** [comment]
No, it's not the receiver and it's not the radio and it's not EdgeTX either. The trace covers the entire processing chain from radio, EdgeTX firmware, ELRS TX module firmware, receiver, ELRS RX firmware and even the ESP32 HW PWM generation.

I have no further conclusion than the entire processing chain is ok. As mentioned earlier your servo might not like the ESP32's 3.3V level for the PWM pulses. If you have a scope available check your servo tester. It might have a 5V level for the PWM pulses.

**Comment by atomGit** [comment]
ps - if you haven't seen the specs for the servo, they're [here](https://www.servocity.com/2000-series-dual-mode-servo-25-2/) - i'd be curious if you might find something that may present a comparability issue

**Comment by atomGit** [comment]
> No, it's not the receiver and it's not the radio and it's not EdgeTX either.

what Tx/Rx combination did you use?

**Comment by mha1** [comment]
The traces where done using a FlySky NV14 on EdgeTX 2.10.2, a BetaFPV Micro TX module on ELRS 3.4.3 and a Radiomaster ER6 on ELRS firmware 3.4.3 too.

**Comment by atomGit** [comment]
perfect - thanks - so everything seems to be pointing to the servos then, agreed?

one last Q: what voltage did you use to power the Rx/servos?

**Comment by mha1** [comment]
![image](https://github.com/user-attachments/assets/db15316f-eb78-44cd-bf6d-cf1daf80f23a)

Well, it shouldn't have a problem with the ESP32's 3.3V level PWM and it has a huge deadband of 4us (4000ns!). The tiny jitter we see in the trace is magnitudes away from causing anything.

**Comment by mha1** [comment]
The receiver was powered by a 2s LiFe battery. Should be about 6.6V. There was no servo connected to the receiver. The measurement was directly taken at the channel 6 servo connector (signal pin).

I can't really offer any other conclusion than everything up to the receivers servo connector is working as I expected it.

**Comment by atomGit** [comment]
> was no servo connected to the receiver

of course - my mistake ... again, thanks - this should be very helpful to John over at ServoCity

**Comment by atomGit** [comment]
quoting myself here...

> i have an old HiTech servo programmer and when i drive one of the servos with that, it works fine, though it's 4.8v whereas i'm normally driving them with 7.4v - i don't know if that makes any difference

**it absolutely does** - feeding the ELRS receiver with 5.1v, the servos run perfectly - i can't test anything between 5.1 and 7.4 at the moment

**Comment by atomGit** [comment]
FINALLY got a definitive answer (from Ethan this time <-- thanks Ethan :1st_place_medal: ) at ServoCity - the newest batch of the [GoBuilda 2000 series servos](https://www.servocity.com/2000-series-dual-mode-servo-25-2/) (and possibly others in that product line ???) do not operate correctly when they are run in continuous mode and the voltage is 7.4, or thereabouts - they do run fine at 5v and 6v though
